### PR TITLE
No more null source positions

### PR DESCRIPTION
**Pull request by monperrus**
It much simplifies client code (no more not-null checks)

### CODE REVIEW

File: src/main/java/spoon/support/reflect/cu/SourcePositionImpl.java [hunk 1]
```diff
@@ -62,53 +62,53 @@ private int searchLineNumber(int[] startLineIndexes, int position) {
...
-		return (position - startLineIndexes[i]) - tabCount + (tabCount * tabSize);
+		return (position - lineSeparatorPositions[i]) - tabCount + (tabCount * tabSize);
 	}
 
 	int[] lineSeparatorPositions;
 
-	private int sourceStart, sourceEnd, line;
+	private int sourceStartName, sourceEnd, sourceStart, line = -1;
 
 	public SourcePositionImpl(CompilationUnit compilationUnit, int sourceStartDeclaration, int sourceStartSource, int sourceEnd, int[] lineSeparatorPositions) {
```
Comment (seintur) [inline src/main/java/spoon/support/reflect/cu/SourcePositionImpl.java]: A bit of Javadocumenting seems welcomed here since the name of the parameters are not really self-intuitive. What is sourceStartDeclaration? What is the difference with sourceStartSource? And why do we need lineSeparatorPositions?

File: src/main/java/spoon/support/reflect/cu/SourcePositionImpl.java [hunk 2]
```diff
@@ -62,53 +62,53 @@ private int searchLineNumber(int[] startLineIndexes, int position) {
...
 		super();
 		this.compilationUnit = compilationUnit;
-		this.sourceStart = sourceStartDeclaration;
+		this.sourceStartName = sourceStartDeclaration;
```
Comment (seintur) [inline src/main/java/spoon/support/reflect/cu/SourcePositionImpl.java]: A bit strange to call a variable with the Name suffix since this is an integer. The semantics of the variable is not easily guessed from its name. Rename?

File: src/main/java/spoon/reflect/visitor/printer/ElementPrinterHelper.java [hunk 3]
```diff
@@ -318,7 +318,7 @@ public void writeComment(CtElement element, CommentOffset offset) {
 			final int sourceEnd = element.getPosition().getSourceEnd();
 			if (offset == CommentOffset.BEFORE && (comment.getPosition().getLine() < line || sourceEnd >= comment.getPosition().getSourceEnd())) {
 				commentsToPrint.add(comment);
-			} else if (offset == CommentOffset.AFTER && comment.getPosition().getSourceStart() >= sourceEnd) {
+			} else if (offset == CommentOffset.AFTER && comment.getPosition().getSourceStart() > sourceEnd) {
```
Comment (tdurieux) [inline src/main/java/spoon/reflect/visitor/printer/ElementPrinterHelper.java]: Why did you change this condition?

File: src/main/java/spoon/support/reflect/cu/SourcePositionImpl.java (hunk 1 above)
Comment (monperrus) [inline src/main/java/spoon/support/reflect/cu/SourcePositionImpl.java]: done

File: src/main/java/spoon/support/reflect/cu/SourcePositionImpl.java (hunk 2 above)
Comment (monperrus) [inline src/main/java/spoon/support/reflect/cu/SourcePositionImpl.java]: done

File: src/main/java/spoon/support/compiler/jdt/PositionBuilder.java [hunk 4]
```diff
@@ -49,7 +50,16 @@ SourcePosition buildPositionCtElement(CtElement e, ASTNode node) {
 		CoreFactory cf = this.jdtTreeBuilder.getFactory().Core();
 		int sourceStartDeclaration = node.sourceStart;
 		int sourceStartSource = node.sourceStart;
-		int sourceEnd = node.sourceEnd;
+		// dzfault value
```
Comment (seintur) [inline src/main/java/spoon/support/compiler/jdt/PositionBuilder.java]: Typo: dzfault -> default

File: src/main/java/spoon/support/reflect/cu/SourcePositionImpl.java [hunk 5]
```diff
@@ -62,53 +62,69 @@ private int searchLineNumber(int[] startLineIndexes, int position) {
...
 		}
-		return (position - startLineIndexes[i]) - tabCount + (tabCount * tabSize);
+		return (position - lineSeparatorPositions[i]) - tabCount + (tabCount * tabSize);
 	}
 
-	int[] lineSeparatorPositions;
+	/** The position of the first byte of this element (incl. documentation and modifiers) */
+	private int sourceStart = -1;
+
+	/** The position of the first byte of this element (incl. documentation and modifiers)  */
```
Comment (seintur) [inline src/main/java/spoon/support/reflect/cu/SourcePositionImpl.java]: The position of the "first" for sourceEnd? Or the "last"?

File: src/main/java/spoon/support/compiler/jdt/PositionBuilder.java (hunk 4 above)
Comment (monperrus) [inline src/main/java/spoon/support/compiler/jdt/PositionBuilder.java]: done

File: src/main/java/spoon/support/reflect/cu/SourcePositionImpl.java (hunk 5 above)
Comment (monperrus) [inline src/main/java/spoon/support/reflect/cu/SourcePositionImpl.java]: done
